docs/config.md
# Configuration options

All the options passed to the executable and visible in the help can also be configured as default in a configuration file.

```sh
jet1090 --help
```

By default, the configuration file is located in:

- `$HOME/.config/jet1090/config.toml` for Linux systems;
- `$HOME/Library/Application\ Support/jet1090/config.toml` for MacOS systems;
- `%HOME%\AppData\Roaming\jet1090\config.toml` for Windows systems

!!! tip "Support for `XDG_CONFIG_HOME`"

    If the `XDG_CONFIG_HOME` variable is set, it takes precedence over the folders detailed above.

    This means you can set this variable and use the `$HOME/.config` folders in MacOS systems as well.

!!! tip

    - You can also set a different configuration file in the `JET1090_CONFIG` environment variable.
    - You can also set that variable in the `.env` (dotenv) file located in the current folder and `jet1090` will look into it.

    If you have several scenarios requiring different configurations files, this option may be where to look at.

## General settings

If you set a configuration file, some parameters must be always present:

```toml
interactive = false      # display a table view
verbose = false          # display decoded messages in the terminal
prevent_sleep = false    # force the laptop not to enter sleep mode (useful when lid is closed)
update_position = false  # auto-update the reference position (useful when on a moving aircraft)
```

Other parameters are optional:

```toml
deduplication = 800        # buffer interval for deduplication, in milliseconds
history_expire = 10        # in minutes
log_file = "-"             # use together with RUSTLOG environment variable
output = "~/output.jsonl"  # the ~ (tilde) character is automatically expanded
redis_url = "redis://localhost:6379"
serve_port = 8080          # for the REST API
```

## Sources

!!! warning

    If you do not want to set any source in the configuration file, you must specify an empty list:

    ```toml
    sources = []
    ```

    Otherwise, **do not** include that line, and set as many sources as you need with the `[[sources]]` header.

### RTL-SDR

The following entry is equivalent to `rtlsdr://serial=00000001@LFBO` except that it sets an alias `rtl-sdr`:

```toml
[[sources]]
name = "rtl-sdr"
rtlsdr = "serial=00000001"
airport = "LFBO"
```

The `airport` parameter replaces the `latitude` and `longitude` parameter if they are not present.

### Beast format

External sources can be configured with the `tcp`, `udp` or `websocket` fields.

```toml
[[sources]]
name = "Toulouse"
tcp = "123.45.67.89:10003"
latitude = 43.5993189
longitude = 1.4362472
```

For the `websocket` you must specify the `ws://` prefix:

```toml
[[sources]]
websocket = "ws://123.45.67.89:8765/zurich"
airport = "LSZH"
```

!!! warning "Reference positions"

    When in a hurry, an airport code is enough to decode [surface messages](https://docs.rs/rs1090/latest/rs1090/decode/bds/bds06/struct.SurfacePosition.html) (otherwise, only `lat_cpr` and `lon_cpr` are provided). It may be useful to fill in precise values for `latitude`, `longitude` and `altitude` for multilateration applications.

!!! warning "Different names for different sources"

    The `name` entry is not mandatory but it is helpful to help recognize different sources in the output format. However, internally, an hashed version of the address is used to uniquely identify sources.

### SeRo Systems

You may input here your [SeRo Systems token](https://doc.sero-systems.de/api/) in order to receive your data. Extra filters are also available in order to limit the network bandwidth.

```toml
[[sources]]
sero.token = ""
sero.df_filter = [17, 18, 20, 21]  # (default: no filter)
# sero.aircraft_filter = []  # list of integer values corresponding to icao24 addresses (default: no filter)
```


changelog.md
# Changelog

## 0.4.2

- Fix BDS 6,5 bug for DF 21
- Implement WebAssembly bindings

## 0.4.1

- Fix building scripts in GitHub releases

## 0.4.0

- Basic deduplication algorithm
- Support for SeRo system data source
- Adjust name aliases and serial numbers
- Major refactoring

## 0.3.9

- New data model
- allow to specify which rtlsdr to use
- fix log_file being ignored in config files
- prefer XDG_CONFIG_HOME if defined over system default, useful in MacOS
- Improve BDS inference based on ground truth messages
- add option to prevent laptop from sleeping
- improve error message when redis server is not reachable

## 0.3.8

- Adapt to Python 3.13
- Propagate errors on non-null CRC chcks

## 0.3.7

- Switch to uv

## 0.3.6

- Select rtl-sdr with arguments
- Typing issue on Redis publish after upgrade

## 0.3.5

- Do not attempt to decode Beast 0x34 messages (avoid errors)
- Fix error forwarding

## 0.3.4

- Major code adjustments following upgrade to deku 0.17.0
- Upgrade cargo-dist, maturin, change version semantics (shared version)

## 0.3.0

- Adjust size for MPSC channels
- Decode Beast format from a websocket source
- Decode RSSI in Beast format, and implement it for RTL-SDR
- Improvement in CPR decoding
- Shell completion
- Support for nix/flake


crates/jet1090/src/cli.rs
use radarcape::BeastSource;
use rs1090::prelude::*;
#[cfg(feature = "rtlsdr")]
use rs1090::source::rtlsdr;
#[cfg(feature = "sero")]
use rs1090::source::sero;
use serde::{Deserialize, Serialize};
use std::collections::hash_map::DefaultHasher;
use std::hash::{Hash, Hasher};
use std::str::FromStr;
use tokio::sync::mpsc::Sender;
use tracing::error;
use url::Url;

// An intermediate structure defined so that you can keep your Sero entries in
// your configuration file even if the sero feature is not activated
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct SeroParams {
    pub token: String,
}

#[cfg(feature = "sero")]
impl From<&SeroParams> for sero::SeroClient {
    fn from(value: &SeroParams) -> Self {
        sero::SeroClient {
            token: value.token.clone(),
        }
    }
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub enum Address {
    Tcp(String),
    Udp(String),
    Websocket(String),
    Rtlsdr(Option<String>),
    Sero(SeroParams),
}

fn build_serial(input: &str) -> u64 {
    // Create a hasher
    let mut hasher = DefaultHasher::new();
    // Hash the string
    input.hash(&mut hasher);
    // Get the hash as a u64
    hasher.finish()
}

impl Source {
    pub fn serial(&self) -> u64 {
        match &self.address {
            Address::Tcp(name) => build_serial(name),
            Address::Udp(name) => build_serial(name),
            Address::Websocket(name) => build_serial(name),
            Address::Rtlsdr(reference) => {
                let name = reference.clone().unwrap_or("rtlsdr".to_string());
                build_serial(&name)
            }
            Address::Sero(_) => 0,
        }
    }
    pub async fn references(&self) -> Vec<(u64, Option<Position>)> {
        match &self.address {
            Address::Tcp(name) => {
                vec![(build_serial(name), self.reference)]
            }
            Address::Udp(name) => {
                vec![(build_serial(name), self.reference)]
            }
            Address::Websocket(name) => {
                vec![(build_serial(name), self.reference)]
            }
            Address::Rtlsdr(reference) => {
                let name = reference.clone().unwrap_or("rtlsdr".to_string());
                vec![(build_serial(&name), self.reference)]
            }
            Address::Sero(params) => {
                #[cfg(feature = "sero")]
                {
                    let sero = sero::SeroClient::from(params);
                    let info = sero.info().await.unwrap();
                    info.sensor_info
                        .iter()
                        .map(|elt| {
                            (
                                elt.sensor.unwrap().serial,
                                elt.gnss.as_ref().unwrap().position.map(
                                    |pos| Position {
                                        latitude: pos.latitude,
                                        longitude: pos.longitude,
                                    },
                                ),
                            )
                        })
                        .collect()
                }
                #[cfg(not(feature = "sero"))]
                {
                    vec![]
                }
            }
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Source {
    address: Address,
    pub name: Option<String>,
    pub reference: Option<Position>,
    #[serde(skip)]
    pub count: u64,
    #[serde(skip)]
    pub last: u64,
}

impl FromStr for Source {
    type Err = String;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        let s = s.replace("@", "?"); // retro-compatibility
        let default_tcp = Url::parse("tcp://").unwrap();

        let url = default_tcp.join(&s).map_err(|e| e.to_string())?;

        let address = match url.scheme() {
            "tcp" => Address::Tcp(format!(
                "{}:{}",
                url.host_str().unwrap_or("0.0.0.0"),
                match url.host() {
                    Some(_) => url.port_or_known_default().unwrap_or(10003),
                    None => {
                        // deals with ":4003?LFBO" (parsed as "tcp:///:4003?LFBO")
                        url.path()
                            .strip_prefix("/:")
                            .unwrap()
                            .parse::<u16>()
                            .expect("A port number was expected")
                    }
                }
            )),
            "udp" => Address::Udp(format!(
                "{}:{}",
                url.host_str().unwrap_or("0.0.0.0"),
                url.port_or_known_default().unwrap()
            )),
            "rtlsdr" => Address::Rtlsdr(url.host_str().map(|s| s.to_string())),
            "ws" => Address::Websocket(format!(
                "ws://{}:{}/{}",
                url.host_str().unwrap_or("0.0.0.0"),
                url.port_or_known_default().unwrap(),
                url.path().strip_prefix("/").unwrap()
            )),
            _ => return Err("unsupported scheme".to_string()),
        };

        let mut source = Source {
            address,
            name: None,
            reference: None,
            count: 0,
            last: 0,
        };

        if let Some(query) = url.query() {
            source.reference = Position::from_str(query).ok()
        };

        Ok(source)
    }
}

impl Source {
    pub async fn receiver(
        &self,
        tx: Sender<TimedMessage>,
        serial: u64,
        name: Option<String>,
    ) {
        match &self.address {
            Address::Rtlsdr(args) => {
                #[cfg(not(feature = "rtlsdr"))]
                {
                    error!("Compile jet1090 with the rtlsdr feature, {:?} argument ignored", args);
                    std::process::exit(127);
                }
                #[cfg(feature = "rtlsdr")]
                {
                    rtlsdr::receiver::<&str>(tx, args.as_deref(), serial, name)
                        .await
                }
            }
            Address::Sero(sero) => {
                #[cfg(not(feature = "sero"))]
                {
                    error!("Compile jet1090 with the sero feature, {:?} argument ignored", sero);
                }
                #[cfg(feature = "sero")]
                {
                    sero::receiver(sero::SeroClient::from(sero), tx).await
                }
            }
            _ => {
                let server_address = match &self.address {
                    Address::Tcp(s) => BeastSource::TCP(s.to_owned()),
                    Address::Udp(s) => BeastSource::UDP(s.to_owned()),
                    Address::Websocket(s) => {
                        BeastSource::Websocket(s.to_owned())
                    }
                    _ => unreachable!(),
                };
                if let Err(e) =
                    radarcape::receiver(server_address, tx, serial, name).await
                {
                    error!("{}", e.to_string());
                }
            }
        }
    }
}

#[cfg(test)]
mod test {
    use super::*;

    #[test]
    fn test_source() {
        let source = Source::from_str("rtlsdr:");
        assert!(source.is_ok());
        if let Ok(Source { address, .. }) = source {
            assert_eq!(address, Address::Rtlsdr(None));
        }

        let source = Source::from_str("rtlsdr://serial=00000001");
        assert!(source.is_ok());
        if let Ok(Source { address, .. }) = source {
            assert_eq!(
                address,
                Address::Rtlsdr(Some("serial=00000001".to_string()))
            );
        }

        let source = Source::from_str("rtlsdr:@LFBO");
        assert!(source.is_ok());
        if let Ok(Source {
            address,
            name,
            reference: Some(pos),
            ..
        }) = source
        {
            assert_eq!(address, Address::Rtlsdr(None));
            assert_eq!(name, None);
            assert_eq!(pos.latitude, 43.628101);
            assert_eq!(pos.longitude, 1.367263);
        }

        let source = Source::from_str("http://default");
        assert!(source.is_err());

        let source = Source::from_str(":4003");
        assert!(source.is_ok());
        if let Ok(Source {
            address: Address::Tcp(path),
            name,
            reference,
            ..
        }) = source
        {
            assert_eq!(path, "0.0.0.0:4003");
            assert_eq!(name, None);
            assert_eq!(reference, None);
        }

        let source = Source::from_str(":4003?LFBO");
        assert!(source.is_ok());
        if let Ok(Source {
            address: Address::Tcp(path),
            name,
            reference: Some(pos),
            ..
        }) = source
        {
            assert_eq!(path, "0.0.0.0:4003");
            assert_eq!(name, None);
            assert_eq!(pos.latitude, 43.628101);
            assert_eq!(pos.longitude, 1.367263);
        }

        let source = Source::from_str("ws://1.2.3.4:4003/get?LFBO");
        assert!(source.is_ok());
        if let Ok(Source {
            address,
            name,
            reference: Some(pos),
            ..
        }) = source
        {
            assert_eq!(
                address,
                Address::Websocket("ws://1.2.3.4:4003/get".to_string())
            );
            assert_eq!(name, None);
            assert_eq!(pos.latitude, 43.628101);
            assert_eq!(pos.longitude, 1.367263);
        }
    }
}
